lib/edgehog_device_forwarder/connections.ex

```diff
@@ -0,0 +1,147 @@
+# Copyright 2023 SECO Mind Srl
+# SPDX-License-Identifier: Apache-2.0
+
+defmodule EdgehogDeviceForwarder.Connections do
+  use Supervisor
+
+  alias EdgehogDeviceForwarder.Connections.TokenData
+
+  @opts Application.compile_env!(:edgehog_device_forwarder, EdgehogDeviceForwarder.Connections)
+  @tokens @opts[:tokens]
+  @messages @opts[:messages]
+
+  @type token :: String.t()
+  @type device_id :: TokenData.device_id()
+  @type realm :: TokenData.realm()
+  @type socket :: TokenData.socket()
+  @type controller :: pid()
+  @type http_request_id :: binary()
+  @type ws_socket_id :: binary()
+
+  # API
+
+  ## Token
+  @spec register_device_info(token, %{:device_id => device_id, :realm => realm}) ::
+          :ok | {:error, :token_not_found}
+  def register_device_info(token, device_info) do
+    ConCache.update(@tokens, token, &TokenData.put_device_info(&1, device_info))
+  end
+
+  @spec register_socket(token, socket) :: :ok | {:error, :already_exists}
+  def register_socket(token, socket) do
+    # TODO: check if we need to handle socket's process termination ourself.
+    # I'd be inclined to say no from
+    #   - https://hexdocs.pm/phoenix/Phoenix.Socket.Transport.html#c:terminate/2
+    #   - https://hexdocs.pm/phoenix/Phoenix.Socket.Transport.html#module-custom-transports
+    ConCache.insert_new(@tokens, token, TokenData.new(socket))
+  end
+
+  @spec close_socket(token) :: :ok
+  def close_socket(token) do
+    ConCache.delete(@tokens, token)
+  end
+
+  @spec fetch_socket(token) :: {:ok, socket} | {:error, :token_not_found}
+  def fetch_socket(token) do
+    case ConCache.get(@tokens, token) do
+      nil -> {:error, :token_not_found}
+      %TokenData{socket: socket} -> {:ok, socket}
+    end
+  end
+
+  ## Messages
+
+  @spec new_http_request(controller) :: {:ok, http_request_id}
+  def new_http_request(controller) do
+    request_id = UUID.uuid4(:raw)
+
+    case ConCache.insert_new(@messages, request_id, {:http, controller}) do
+      :ok -> {:ok, request_id}
+      {:error, :already_exists} -> new_http_request(controller)
```

Comment:
Is that infinite recursion or am I missing something?

Reply:
it stops when it generates a new UUIDv4 (not already in the cache)